2week: implement Lotto Simulation Console Program

leeheefull opened the pull request.

leeheefull committed:
2week: lotto model 개발

leeheefull committed:
2week: mvc pattern lotto simulation 개발
- model 개발
- LottoModule 개발
- LottoController 개발
- LottoView 개발

leeheefull committed:
2week: before first code review

leeheefull committed:
2week: after first code review
- 접근지정자 수정
- LottoModule -> LottoGame 명칭 변경
- 자료구조: ArrayList -> TreeSet
- 함수: whatIsRanking -> getRanking
- boolean method 수정
- object 출력의 위험성 제거

leeheefull committed:
2week: today i learnd
- Stream 문법

leeheefull committed:
2week: after second code review
- 주석 제거
- 접근지시제어자 사용
- Lotto: abstact 제거, public 제거
- LottoController: 출력문을 위한 for문 view로 보내기
- LottoGame: setLottoListResult(): if-else 수정
- LottoGame: clear()로 생명주기 만들기
- util: LottoConstants: Menu 만들기

leeheefull committed:
2week: today i learnd
- Bounded Type Parameters

leeheefull committed:
2week: WinningResult: use enum

leeheefull committed:
2week: update READM.md

jeonyeonbin commented on 2week/LottoSimulation/src/kr/ac/hs/se/model/WinningResult.java:
```diff
@@ -0,0 +1,70 @@
+            return FOUR;
+        }
+    },
+    SIX(6, "낙첨") {
```
SIX SEVEN EIGHT을 명시적으로 적어주는것도 좋을수있겟지만,
OTHERS라고 나머지들은 낙첨으로 주시는것도 나쁘지 않을것같아요.

leeheefull replied:
좋은 방법입니다! 수정했습니다

jeonyeonbin commented on 2week/LottoSimulation/src/kr/ac/hs/se/model/WinningLotto.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Override
+    public void setBasicNumbers() {
+        while (this.basicNumbers.size() < SIZE) {
```
basicNumbers에 대해서 해당 object내부에있으면 TC를 짜기가 곤란하다 라는 리뷰를 드렸었는데,
혹시 그 리뷰는 아직 반영이 안되걸까요?

leeheefull replied:
```java
public class LottoSample {

    private static Set<Integer> randomNumbers;

    public LottoSample() {
        randomNumbers = new TreeSet<>();
    }

    public Set<Integer> createRandomNumbers() {
        while (randomNumbers.size() < 6) {
            int tmp = (int) (Math.random() * 42) + 1;
            randomNumbers.add(tmp);
        }
        return randomNumbers;
    }

    public int createRandomNumber() {
        int tmp = (int) (Math.random() * 42) + 1;
        while (randomNumbers.contains(tmp)) {
            tmp = (int) (Math.random() * 42) + 1;
        }
        return tmp;
    }
}
```

TC를 위한 LottoSample을 해주는 class를 만들어 수정했습니다!

jeonyeonbin replied:
Sample 코드를 만들라는 이야기는 아니였습니다.

RandomGenrator라는 Util class를 만들어서 그 값에서 6개의 숫자를 랜덤하게 뱉어낼수있는 코드를 주고 Lotto의 경우에는 그저 setter만 태울수있게 하라는 이야기였습니다.

Test를 위한 코드가 business코드에는 존재하면 안됩니다.

leeheefull replied:
제가 클래스 이름을 잘못 지었습니다.
연빈님이 말씀하신 내용은 이해됐습니다
위의 코드로 setter에 태울 예정이었습니다

jeonyeonbin commented on 2week/LottoSimulation/src/kr/ac/hs/se/game/LottoGame.java:
```diff
@@ -0,0 +1,71 @@
+
+    public void clearLottoGame() {
+        for (PurchasedLotto purchasedLotto : purchasedLottoList) {
+            purchasedLotto.clear();
```
for문으로 clear하고 다음라인의 List에서도 클리어를 하는데 이유가있을까요?

leeheefull replied:
완벽하게 삭제를 해야겠단 생각을 했었는데
다시보니 의미가 없는 것 같습니다! 수정했습니다

jeonyeonbin commented on 2week/LottoSimulation/src/kr/ac/hs/se/game/LottoGame.java:
```diff
@@ -0,0 +1,71 @@
+
+    public void setLottoListResult() {
+        for (PurchasedLotto purchasedLotto : purchasedLottoList) {
+            setRank(purchasedLotto);
```
위와 같은코드가 coupling이 강하게 생길수있는 코드입니다.

purchasedLotto가 다른메소드를 타고왓는데 내부의 값이 변경이 되어있는 상황이죠.

이런 상황에는 어떻게 하는게 더 좋은 코드를 만들 수 있을까요??

leeheefull replied:
지적해주신 덕분에 "coupling은 낮추고 cohesion는 높여라"라는 말을 알게되었습니다.
coupling이 강해지지 않도록 코드를 수정했습니다.

jeonyeonbin commented on 2week/LottoSimulation/src/kr/ac/hs/se/model/WinningResult.java:
```diff
@@ -0,0 +1,70 @@
+        return result;
+    }
+
+    public abstract WinningResult getNextRank();
```
다음번 enum을 이용한 abstract method를 만들어야될땐, FunctionalInterface 중 Function<I, O>, TriFunction<T, F, R> 

위와같은 functional interface를 이용해서 한번 만들어보시고 

functional interface가 java에서 무슨역할을하고 어떻게 되는지 한번 보시면 좋을 것 같아요.

leeheefull replied:
넵 알겠습니다!

jeonyeonbin commented on 2week/LottoSimulation/src/kr/ac/hs/se/controller/LottoController.java:
```diff
@@ -0,0 +1,73 @@
+
+    private void createLotto(int numberOfLotto) {
+        lottoGame.setWinningLotto();
+        lottoGame.setPurchasedLottoList(numberOfLotto);
```
lottoGame에 setting을하고 purchaseLottoList를 setting하는걸 외부에서 알아야 될지 한번 고민해보시는것도 좋을것같아요.

start라는 method로 명시적으로 start가 되는게 좋지않을까요??

leeheefull replied:
넵 수정했습니다!

jeonyeonbin commented on 2week/LottoSimulation/src/kr/ac/hs/se/game/LottoGame.java:
```diff
@@ -0,0 +1,71 @@
+            }
+        }
+        int bonusNumber = this.winningLotto.getBonusNumber();
+        if (purchasedLotto.getWinningResult().getRank() == 2 && isWinningBonusNumber(purchasedLotto, bonusNumber)) {
```
2라는 숫자의 Magic Number가 아닌 enum을 활용하셧으니 enum끼리 비교해도 되어보이는데요

leeheefull replied:
맞는 말씀입니다 수정했습니다!

leeheefull committed:
2week: today i learned
- Coupling, Cohesion
- Java naming conventions
- Functioal Interface
- Lombok

leeheefull committed:
2week: after last code review
- WinningResult: SIX, SEVEN, EIGHT -> OTHERS
- LottoSample
- clear 수정
- LottoGame: coupling 해결
- LottoGame: enum사용
